### PR TITLE
Uri parse userinfo

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

*Description of changes:*
- Add parsing of userinfo for our URI structure.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/uri.c [hunk 1]
```diff
@@ -327,18 +327,39 @@ static void s_parse_authority(struct uri_parser *parser, struct aws_byte_cursor
...
+        if (userinfo_delim) {
+
+            parser->uri->userinfo =
+                aws_byte_cursor_advance(&authority_parse_csr, userinfo_delim - authority_parse_csr.ptr);
+            /* For the "@" mark */
+            aws_byte_cursor_advance(&authority_parse_csr, 1);
+            struct aws_byte_cursor userinfo_parse_csr = parser->uri->userinfo;
+            uint8_t *info_delim = memchr(userinfo_parse_csr.ptr, ':', userinfo_parse_csr.len);
+            /* RFC-3986 section 3.2.1: Use of the format "user:password" in the userinfo field is deprecated. But still,
+             * we will use this format here as it's a common use case. You also have the whole userinfo field. */
```
Comment (TingDaoK) [inline source/uri.c:341]: TBD, I don't know what's the better way... Or we get a list of cursor for `userinfo = *( unreserved / pct-encoded / sub-delims / ":" )`? I feel like this way is better, but open to discussion.😃

Comment (TingDaoK) [reply source/uri.c:341]: https://datatracker.ietf.org/doc/html/rfc1738#section-3.1 hmmm

File: source/uri.c [hunk 2]
```diff
@@ -327,18 +327,41 @@ static void s_parse_authority(struct uri_parser *parser, struct aws_byte_cursor
...
+        if (userinfo_delim) {
+
+            parser->uri->userinfo =
+                aws_byte_cursor_advance(&authority_parse_csr, userinfo_delim - authority_parse_csr.ptr);
+            /* For the "@" mark */
+            aws_byte_cursor_advance(&authority_parse_csr, 1);
+            struct aws_byte_cursor userinfo_parse_csr = parser->uri->userinfo;
+            uint8_t *info_delim = memchr(userinfo_parse_csr.ptr, ':', userinfo_parse_csr.len);
+            /* RFC-3986 section 3.2.1: Use of the format "user:password" in the userinfo field is deprecated. But we
+             * treat the userinfo as URL here, also, if the format is not following URL partern, you have the hole
```
Comment (justinboswell) [inline source/uri.c:341]: whole

**Review by justinboswell — APPROVED** [review]

File: source/uri.c (hunk 2 above)
Comment (TingDaoK) [reply source/uri.c]: aha, you-got-me 🙆‍♂️

File: source/uri.c [hunk 3]
```diff
@@ -327,18 +327,41 @@ static void s_parse_authority(struct uri_parser *parser, struct aws_byte_cursor
...
+        if (userinfo_delim) {
+
+            parser->uri->userinfo =
+                aws_byte_cursor_advance(&authority_parse_csr, userinfo_delim - authority_parse_csr.ptr);
+            /* For the "@" mark */
+            aws_byte_cursor_advance(&authority_parse_csr, 1);
+            struct aws_byte_cursor userinfo_parse_csr = parser->uri->userinfo;
+            uint8_t *info_delim = memchr(userinfo_parse_csr.ptr, ':', userinfo_parse_csr.len);
+            /* RFC-3986 section 3.2.1: Use of the format "user:password" in the userinfo field is deprecated. But we
+             * treat the userinfo as URL here, also, if the format is not following URL partern, you have the whole
```
Comment (bretambrose) [inline source/uri.c:341]: pattern

**Review by bretambrose — APPROVED** [review]

**Pull request merged**